### PR TITLE
Avoid buffer copies in netstring

### PR DESCRIPTION
**Pull request by mhofman**
Rewrite the reader and writer of the netstring stream transform to avoid buffer copies and allocations as much as possible.
- The reader delays allocation of a buffer until the length prefix is parsed.
- The reader re-uses the input buffer if the whole data is a subset of the incoming chunk
  - possible point of improvement is to handle the trailing separator (comma) in a separate chunk, as may happen went sent chunked in the first place
- The writer can output in chunked mode, where the prefix, data and separator are sent to the output as separate writes/next calls.
- The writer can handle an array of buffers as input for producers which would otherwise concatenate their data.
- The writer no longer re-uses an internal buffer for its output as a consumer that doesn't fully process the buffer (e.g. caches it) before resolving would get corrupted data. While Node.js stream sinks should be safe since our adapter doesn't resolve until drain / write completion, not all sinks may be as robust.

Fix a bug discovered in node-stream where the writer's `next` didn't return an iterator result.

### CODE REVIEW

**Comment by mhofman** [comment]
Best reviewed commit-by-commit

File: packages/stream-node/writer.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ export const makeNodeWriter = writer => {
           } else {
             resolve(undefined);
           }
-        }),
+        }).then(() => ({ done: false, value: undefined })),
```
Comment (kriskowal) [inline packages/stream-node/writer.js:60]: We could harden and intern the medial iteration result.

We could also avoid some promise and closure allocations by changing `resolve(undefined)` to `resolve(nonFinalIteration)` and creating a `() => resolve(nonFinalIteration)` closure only if `writer.write` is false.

File: packages/netstring/writer.js [hunk 2]
```diff
@@ -1,51 +1,54 @@
...
+      const prefix = getLengthPrefixCharCodes(messageLength);
 
-      buffer[colonAt] = COLON;
-      buffer.set(message, messageAt);
-      buffer[commaAt] = COMMA;
+      if (chunked) {
+        return Promise.all([
+          output.next(new Uint8Array(prefix)),
+          ...messageChunks.map(async chunk => output.next(chunk)),
+          output.next(COMMA_BUFFER),
```
Comment (kriskowal) [inline packages/netstring/writer.js:35]: I’m not sure whether calling `next` after a final iteration has well-defined behavior. I don’t think this is wrong, but it might be good to document somewhere that this writer might write after the target closes.

File: packages/netstring/writer.js [hunk 3]
```diff
@@ -1,51 +1,54 @@
...
-// It also must be long enough to fit an ASCII decimal representation of the
-// length of the longest possible message.
-// This is a guess:
-const initialBufferLength = 128;
-
-const encoder = new TextEncoder();
+/** @param {number} length */
+const getLengthPrefixCharCodes = length =>
+  // eslint-disable-next-line no-bitwise
+  [...`${length | 0}:`].map(char => char.charCodeAt(0));
```
Comment (kriskowal) [inline packages/netstring/writer.js:9]: Switching from `map` to a `for` loop may yield a measurable improvement.
```suggestion
const getLengthPrefixCharCodes = length => {
  // eslint-disable-next-line no-bitwise
  const text = String(length | 0);
  const codes = new Uint8Array(text.length + 1);
  let i;
  for (i = 0; i < codes.length; i += 1) {
    codes[i] = text.charCodeAt(i);
  }
  codes[i] = COLON;
}
```
Can also avoid the intermediate Array and go straight to Uint8Array.

File: packages/netstring/writer.js [hunk 4]
```diff
@@ -1,51 +1,54 @@
...
-        buffer,
+    async next(messageChunks) {
+      if (!Array.isArray(messageChunks)) {
+        messageChunks = [messageChunks];
+      }
+
+      const messageLength = messageChunks.reduce(
+        (acc, { length }) => acc + length,
+        0,
       );
```
Comment (kriskowal) [inline packages/netstring/writer.js:27]: This could have been a for loop.

File: packages/netstring/test/test-netstring.js [hunk 5]
```diff
@@ -20,81 +20,115 @@ async function read(source) {
...
+test('fails reading no colon', readErroneousChunkedMessage, ['1A,']);
 
 function delay(ms) {
   return new Promise(resolve => {
     setTimeout(resolve, ms);
   });
 }
 
-const makeArrayWriter = () => {
+const makeArrayWriter = opts => {
```
Comment (kriskowal) [inline packages/netstring/test/test-netstring.js:106]: Very ignorable nit: does our style guide have anything to say about abbreviations? My personal preference is to spell these out.

File: packages/netstring/test/test-netstring.js [hunk 6]
```diff
@@ -119,9 +155,25 @@ test('concurrent writes', async t => {
...
-});
+};
+test('concurrent writes', concurrentWrites);
+test('concurrent writes (chunked)', concurrentWrites, { chunked: true });
 
-test('round-trip varying messages', async t => {
+const chunkedWrite = async (t, opts) => {
+  const { array, writer } = makeArrayWriter(opts);
+  const strChunks = ['hello', ' ', 'world'];
+  await writer.next(strChunks.map(strChunk => encoder.encode(strChunk)));
```
Comment (kriskowal) [inline packages/netstring/test/test-netstring.js:165]: 👀 Apparent signature change from bytes to array of byte arrays.

File: packages/netstring/reader.js [hunk 7]
```diff
@@ -2,67 +2,113 @@
...
  * @param {string} [opts.name]
- * @param {number} [opts.capacity]
  */
-async function* makeNetstringIterator(
-  input,
-  { name = '<unknown>', capacity = 1024 } = {},
-) {
-  let length = 0;
-  let buffer = new Uint8Array(capacity);
+async function* makeNetstringIterator(input, { name = '<unknown>' } = {}) {
```
Comment (kriskowal) [inline packages/netstring/reader.js:14]: I’ve heard that generators are slow and that state machines are the way to go for speed. I used a generator in hopes that a the readability would be more important than speed. This appears to be the time to revisit that choice. But, I like testing the fast one in comparison to the slow-but-readable one for correctness, especially in a fuzzer.

File: packages/netstring/reader.js [hunk 8]
```diff
@@ -2,67 +2,113 @@
...
+        let i = 0;
+        while (i < buffer.length) {
+          const c = buffer[i];
+          i += 1;
+          if (c >= ZERO && c <= NINE) {
+            lengthBuffer.push(c);
+          } else if (c === COLON && lengthBuffer.length) {
+            lengthBuffer.push(c);
+            break;
+          } else {
```
Comment (kriskowal) [inline packages/netstring/reader.js:47]: Should throw an error of the lengthBuffer length exceeds the maximum length necessary to express the maximum message length, so the reader can’t be induced to OOM by a long length prefix.

Comment (kriskowal) [reply packages/netstring/reader.js:47]: Similarly below. Can punt to an issue since it would be a new level of hardening unrelated to this performance push.

File: packages/netstring/reader.js [hunk 9]
```diff
@@ -2,67 +2,113 @@
...
+          if (Number.isNaN(remainingDataLength)) {
+            throw new Error(
+              `Invalid netstring prefix length ${prefix} at offset ${offset} of ${name}`,
+            );
+          }
+          offset += lengthBuffer.length + 1;
+          lengthBuffer = null;
+        }
+      }
+
```
Comment (kriskowal) [inline packages/netstring/reader.js:71]: I unfortunately need to lean heavily on tests to verify the logic in here. I don’t see any flaws but from experience writing similar code, I’m sure that’s no indication. A fuzzer would make me more comfortable. A boundary searcher would be even better. A boundary searcher would be a test that visits various permutations of message sizes +/-1 around powers of two or other likely boundary cases, in various orders, trying to tease out edge cases.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Good call on making {chunked} writing opt-in.

We need:

- [ ] fuzzer (a test that runs a producer and consumer using a PRNG to generate a very long sequence of messages)
- [ ] benchmark (before and after figures)
- [ ] confirmation that agoric-sdk still passes with this dependency pinned

 I could be convinced to run with scissors except for the last bit.

**Comment by kriskowal** [comment]
Good catch re the undefined final iteration result!

**Comment by mhofman** [comment]
> * [ ]  fuzzer (a test that runs a producer and consumer using a PRNG to generate a very long sequence of messages)
Doesn't one of the unit test already do this?
> * [ ]  benchmark (before and after figures)
I tried to get perf before/after in agoric-sdk, but the end-to-end timings are very noisy. Didn't have time to get better data
> * [ ]  confirmation that agoric-sdk still passes with this dependency pinned
I did run agoric-sdk tests including integration with this PR as a patch-package

File: packages/stream-node/writer.js (hunk 1 above)
Comment (mhofman) [reply packages/stream-node/writer.js:60]: Yeah I considered those as well, but went for this for simplicity's sake. If you have a preference I'm happy to go any other way.

File: packages/netstring/writer.js (hunk 4 above)
Comment (mhofman) [reply packages/netstring/writer.js:27]: A simple accumulator is the one case where I feel like `reduce` is acceptable but if it hurts too much to read, I'm happy to make it a loop.

File: packages/netstring/writer.js (hunk 3 above)
Comment (mhofman) [reply packages/netstring/writer.js:9]: Given the size of the iteration I don't think the performance impact is measurable. Plus v8 can optimize these quite a bit these days.

I also wanted to avoid allocating an `ArrayBuffer` in the case where I just need to set the charCodes in an existing buffer. Not sure what's better, a temporary `ArrayBuffer` + `UInt8Array` or a temporary `Array`

File: packages/netstring/reader.js (hunk 7 above)
Comment (mhofman) [reply packages/netstring/reader.js:14]: Yeah since we already have a state machine inside the generator, you're right going all the way might be better.

File: packages/netstring/reader.js (hunk 8 above)
Comment (mhofman) [reply packages/netstring/reader.js:47]: What do you mean by maximum message length? I believe the netstring spec doesn't have such a notion (even though the example implies a max length of 999999999 bytes). The tagged netstring spec does have that length as specified limit.

Would we consider 1GiB-1 as the upper limit, or would it be a config option of the reader?

File: packages/netstring/reader.js (hunk 9 above)
Comment (mhofman) [reply packages/netstring/reader.js:71]: I'm not sure what powers of 2 have to do with the length prefix logic here. I believe the unit tests already cover most edge cases, which are:
- break in the middle of the prefix
- break between the number prefix and the colon
- invalid character in the prefix

The one case not covered is an empty prefix before the colon, and I can definitely fix that.

Can you think of any other cases to test for?

File: packages/netstring/test/test-netstring.js (hunk 5 above)
Comment (mhofman) [reply packages/netstring/test/test-netstring.js:106]: Arguably I also prefer spelling out but used the term `opts` because I copied it from the JSDoc `@param` of `makeNetstringWriter`, which I itself copied from the existing reader one.

File: packages/netstring/test/test-netstring.js (hunk 6 above)
Comment (mhofman) [reply packages/netstring/test/test-netstring.js:165]: Correct, `chunkedWrite` tests the new chunks receiving mode.

File: packages/netstring/writer.js (hunk 2 above)
Comment (mhofman) [reply packages/netstring/writer.js:35]: Yeah it's definitely an aggressive writer in chunk writing mode. Good thing it's not the default, and the node-stream writer can take it. I can add a note.

File: packages/netstring/test/test-netstring.js (hunk 5 above)
Comment (kriskowal) [reply packages/netstring/test/test-netstring.js:106]: _mea culpa_!

File: packages/netstring/reader.js (hunk 8 above)
Comment (kriskowal) [reply packages/netstring/reader.js:47]: We should definitely have a config option and highly encourage a reasonable limit. Protobuf has a hard limit at 2GB, which is close enough to Infinity to be a reasonable default but not adequate for limiting the sender’s ability to induce an OOMkill remotely.

**Review by kriskowal — APPROVED** [review]
- [x] round-trip messages of varying lengths is good enough to cover the fuzzer case. Thanks for reminding me of its existence. I grepped for fuzz.
- [x] I’m satisfied that you got an end-to-end test passing
- [ ] Would be nice to have a benchmark to confirm this moves performance in the right direction in isolation. Benchmarks are surprising. I’m imagining something lightweight, like a generalization of the round-trip test that runs until some minimum time threshold and reports the iteration count over the total time.

File: packages/netstring/reader.js (hunk 7 above)
Comment (mhofman) [reply packages/netstring/reader.js]: Decided to not do as consuming the input without a `for await of` loop makes things more complicated.

File: packages/netstring/reader.js (hunk 8 above)
Comment (mhofman) [reply packages/netstring/reader.js:61]: Done, made configurable with the default being what the netstring specs imply.

File: packages/netstring/reader.js (hunk 9 above)
Comment (mhofman) [reply packages/netstring/reader.js:89]: Added an empty prefix test

File: packages/netstring/writer.js (hunk 2 above)
Comment (mhofman) [reply packages/netstring/writer.js:45]: Note added in the description of the maker.

File: packages/stream-node/writer.js (hunk 1 above)
Comment (mhofman) [reply packages/stream-node/writer.js]: Fixed. Also hardened the `finalIteration` result as well.

**Review by mhofman — COMMENTED** [review]
@kriskowal PTAL

**Review by kriskowal — APPROVED** [review]
